### ISSUE TITLE
[lld/mac] LTO doesn't support weak symbols

### ISSUE THREAD

**Issue by nico** [llvm/llvm-project#52778]
Let's make a dylib, imagine this being a system dylib (imagine that `foo` would only exist in the 11.0 version of the dylib, but the dylib exists on 10.15 too -- just not foo() in it).

```
% cat foo.c                                               
void foo() {}
% clang -shared -o foo.dylib foo.c -mmacosx-version-min=10.15
```

Let's make a program that links against it:

```
% cat avail.c 
__attribute__((availability(macos, introduced=11.0))) void foo(void);
int main() {
  if (__builtin_available(macos 11.0, *))
    foo();
}
% out/gn/bin/clang avail.c foo.dylib -mmacosx-version-min=10.15 -isysroot $(xcrun -show-sdk-path) -fuse-ld=lld 
```

`_foo` is referenced as a weak external:

```
% nm -m a.out | grep _foo
                 (undefined) weak external _foo (from foo)
```

Now let's do the same with lto:

```
% out/gn/bin/clang avail.c foo.dylib -mmacosx-version-min=10.15 -isysroot $(xcrun -show-sdk-path) -fuse-ld=lld -flto=thin
% nm -m a.out | grep _foo
                 (undefined) external _foo (from foo)
```

(Extracted from https://bugs.chromium.org/p/chromium/issues/detail?id=1280923#c6 upstream – this made Google Chrome Canary crash on macOS 10.15 or older after we turned on LTO.)

**Comment by nico** [comment]
In lld/MachO/InputFiles.cpp

```
static macho::Symbol *createBitcodeSymbol(const lto::InputFile::Symbol &objSym,
                                          BitcodeFile &file) {
  StringRef name = saver.save(objSym.getName());

  // TODO: support weak references
```

Well, sounds like it's time to do that TODO :) @int3, you wrote that TODO – do you happen to have an idea what needs to be done for that TODO?

**Comment by nico** [comment]
Maybe this is all that's needed?

```diff
-  // TODO: support weak references
   if (objSym.isUndefined())
-    return symtab->addUndefined(name, &file, /*isWeakRef=*/false);
+    return symtab->addUndefined(name, &file, /*isWeakRef=*/objSym.isWeak());
```

Need to check ld64 if just using the `isWeak()` bit on unrefs is supposed to mean "weak ref" instead of "weak def".

**Comment by nico** [comment]
https://github.com/apple-opensource/ld64/blob/e28c028b20af187a16a7161d89e91868a450cadc/src/ld/parsers/lto_file.cpp#L469

https://github.com/llvm/llvm-project/blob/e356027016c6365b3d8924f54c33e2c63d931492/llvm/tools/lto/lto.cpp#L316
https://github.com/llvm/llvm-project/blob/e356027016c6365b3d8924f54c33e2c63d931492/llvm/include/llvm/LTO/legacy/LTOModule.h#L136

https://github.com/llvm/llvm-project/blob/89b57061f7b769e9ea9bf6ed686e284f3e55affe/llvm/lib/LTO/LTOModule.cpp#L562

So looks like ld64 bases it on hasExternalWeakLinkage() on the GV in the end.

lld gets it from https://github.com/llvm/llvm-project/blob/89b57061f7b769e9ea9bf6ed686e284f3e55affe/llvm/include/llvm/Object/IRSymtab.h#L199 which is written here https://github.com/llvm/llvm-project/blob/89b57061f7b769e9ea9bf6ed686e284f3e55affe/llvm/include/llvm/Object/IRSymtab.h#L325 / https://github.com/llvm/llvm-project/blob/89b57061f7b769e9ea9bf6ed686e284f3e55affe/llvm/include/llvm/Object/IRSymtab.h#L102

Related reading: https://reviews.llvm.org/D31364

**Comment by nico** [comment]
http://llvm-cs.pcc.me.uk/lib/Bitcode/Writer/BitcodeWriter.cpp#4401 and its refs too.

**Comment by nico** [comment]
Aha, http://llvm-cs.pcc.me.uk/lib/Object/ModuleSymbolTable.cpp#184 / http://llvm-cs.pcc.me.uk/lib/Object/IRSymtab.cpp#223 is what sets the irsymtab flags based on GlobalValues.

github links to same (for inline preview):

https://github.com/llvm/llvm-project/blob/08ed216/llvm/lib/Object/ModuleSymbolTable.cpp#L192

https://github.com/llvm/llvm-project/blob/ac02bca/llvm/lib/Object/IRSymtab.cpp#L241

**Comment by nico** [comment]
Well it's better than the current behavior for sure, so https://reviews.llvm.org/D115949 :)

**Comment by int3** [comment]
> Well it's better than the current behavior for sure

Haven't had time to read through all the linked code yet, but why this caveat? What else might we be missing?

**Comment by nico** [comment]
1. I tried to trace through the code to make sure we use the same bits as ld64, and I got a bit lost along the way. I think it might not be 100% the same condition we use in my patch (but it's possible that for Undefineds, it's equivalent).
2. I haven't checked yet that Chrome LTOd with my patch actually works.

It's possible everything's good with the patch (not verified yet because only an arm mac with macOS 12 at hand atm), and it's certainly a progression over the current state :)

**Comment by int3** [comment]
Gotcha. Maybe add that to the commit message?

**Comment by nico** [comment]
(Done.)

**Comment by nico** [comment]
I've now confirmed that this does indeed fix the startup crash on my 10.15 intel mbp. LTO'd Chromium starts up fine and seems to work now.

I made a standalone repro archive with LTO inputs since that seems like something useful to have. It's here: https://drive.google.com/file/d/1tmCbXWu32jFL09_Xog8dIW6TczuK4AaK/view?usp=sharing

---

To make the repro:

* at chromium rev 73f3a9a9380e239fbdc6887c660a30bb1db5d3ca
* (slightly unusual setup in that I was doing an arm->intel cross build and I had to slightly hack up my local environment; https://crbug.com/1280968 comment 1 has all local changes linked)

args.gn:

```
is_debug = false
is_official_build = true
enable_dsyms = false
target_cpu = "x64"
use_goma = true
symbol_level = 0
```

I ran:

1. `time ninja -C out/gn -j250 chrome`
2. `rm out/gn/obj/chrome/chrome_framework_shared_library/Chromium\ Framework`
3. `time ninja -C out/gn -j250 chrome -v -d keeprsp`
4. Copy the link command for the main framework
5. `cd out/gn`, paste link command, add `-Wl,--reproduce=repro.tar` at the end
6. Copy repro.tar into a different directory, untar again
7. Copy `out/gn/Chromium.app` into the directory where it was untarred, next to response.txt
8. Edit response.txt so that the `-o` flag writes Chromium.framework into the app (`-o "Chromium.app/Contents/Frameworks/Chromium Framework.framework/Versions/99.0.4770.0/Chromium Framework"` in this case, but the version number is different each time)
9. `rm 'Chromium.app/Contents/Frameworks/Chromium Framework.framework/Versions/99.0.4770.0/Chromium Framework'` to make the file a tiny bit smaller
10. tar up again: `time tar czf repro.tbz repro`
